### ISSUE TITLE
Error when update a null value in Postgres

### ISSUE THREAD

**Issue by GabrielRozendo** [simolus3/moor#2441]
I was in drift 2.5.0, without the `postgres_v3_experimental`, holding on to upgrade but I did to 2.8.0 last week and since then a problem started wherever I update to null a `integer().nullable()` column with `Value(null)`.

I've just simulated in a fresh project right now.

## Error

> PostgreSQLSeverity.error 42804: column "a_int_nullable_column" is of type bigint but expression is of type text Hint: You will need to rewrite or cast the expression.

### Steps to reproduce:

1. Docker with a fresh Postgres database, `image: postgres:15`
2. Create a new dart console project 
3. Followed the example in [extras/drift_postgres](https://github.com/simolus3/drift/blob/develop/extras/drift_postgres/example/test.dart) repo with `Todos` table from [Drift in pub example](https://pub.dev/packages/drift/example). (Full code below...)
6. `dart run build_runner build`
7. `dart run bin/main.dart`

### Version

- Dart SDK version: 3.0.1 (stable)
- PostgreSQL 15.2
- Drift 2.8.0

### Code

pubspec.yaml
```yaml
dependencies:
  drift: ^2.8.0
  drift_postgres:
    git:
      url: https://github.com/simolus3/drift.git
      path: extras/drift_postgres

dev_dependencies:
  build_runner: ^2.4.4
  drift_dev: ^2.8.2
```

my_database.dart
```dart
import 'package:drift/drift.dart';
part 'my_database.g.dart';

class Todos extends Table {
  IntColumn get id => integer().autoIncrement()();
  TextColumn get titleNullableColumn => text().nullable()();
  IntColumn get aIntNullableColumn => integer().nullable()();
}

@DriftDatabase(tables: [Todos])
class MyDatabase extends _$MyDatabase {
  MyDatabase(super.e);

  @override
  int get schemaVersion => 1;
}

```

main.dart
```dart
import 'package:drift/drift.dart';
import 'package:drift_postgres/postgres.dart';
import 'package:postgres/postgres_v3_experimental.dart';

import 'my_database.dart';

Future<void> main(List<String> arguments) async {
  final postgres = PgDatabase(
    endpoint: PgEndpoint(
      host: 'localhost',
      database: 'docker',
      username: 'postgres',
      password: 'postgres',
    ),
    logStatements: true,
  );

  final db = MyDatabase(postgres);
  final id = await db.into(db.todos).insert(
        TodosCompanion.insert(
          titleNullableColumn: Value('my title'),
          aIntNullableColumn: Value(1),
        ),
      );

  final query0 = db.update(db.todos)..where((t) => t.id.equals(id));
  final rows0 = await query0.write(
    TodosCompanion(titleNullableColumn: Value(null)),
  );
  print(rows0);

  final query1 = db.update(db.todos)..where((t) => t.id.equals(id));
  final rows1 = await query1.write(
    TodosCompanion(aIntNullableColumn: Value(2)),
  );
  print(rows1);

  final query2 = db.update(db.todos)..where((t) => t.id.equals(id));
  final rows2 = await query2.write(
    TodosCompanion(aIntNullableColumn: Value(null)), // throws the error
  );
  print(rows2);
}

```

`dart run bin/main.dart`

### Output

> Drift: Sent CREATE TABLE IF NOT EXISTS "todos" ("id" bigserial PRIMARY KEY NOT NULL, "title_nullable_column" text NULL, "a_int_nullable_column" bigint NULL); with args []
> Drift: Sent INSERT INTO "todos" ("title_nullable_column", "a_int_nullable_column") VALUES ($1, $2) RETURNING id with args [my title, 1]
> **Drift: Sent UPDATE "todos" SET "title_nullable_column" = $1 WHERE "id" = $2; with args [null, 1]**
> 1
> **Drift: Sent UPDATE "todos" SET "a_int_nullable_column" = $1 WHERE "id" = $2; with args [2, 1]**
> 1
> **Drift: Sent UPDATE "todos" SET "a_int_nullable_column" = $1 WHERE "id" = $2; with args [null, 1]**
> Unhandled exception:
> **PostgreSQLSeverity.error 42804: column "a_int_nullable_column" is of type bigint but expression is of type text Hint: You will need to rewrite or cast the expression.** 
> #0      PgConnectionImplementation._sendAndWaitForQuery (package:postgres/src/v3/connection.dart:228:30)
> #1      PgConnectionImplementation.prepare (package:postgres/src/v3/connection.dart:266:11)
> #2      PgConnectionImplementation.execute (package:postgres/src/v3/connection.dart:284:30)
> #3      _PgDelegate._runWithArgs (package:drift_postgres/src/pg_database.dart:93:34)
> #4      _PgDelegate.runUpdate (package:drift_postgres/src/pg_database.dart:116:12)
> #5      _BaseExecutor.runUpdate.<anonymous closure> (package:drift/src/runtime/executor/helpers/engines.dart:87:19)
> #6      _BaseExecutor._synchronized.<anonymous closure> (package:drift/src/runtime/executor/helpers/engines.dart:57:22)
> #7      new Future.sync (dart:async/future.dart:305:31)
> #8      Lock.synchronized.callBlockAndComplete (package:drift/src/utils/synchronized.dart:17:21)
> #9      Lock.synchronized.<anonymous closure> (package:drift/src/utils/synchronized.dart:21:35)
> <asynchronous suspension>
> #10     UpdateStatement._performQuery.<anonymous closure> (package:drift/src/runtime/query_builder/statements/update.dart:37:14)
> <asynchronous suspension>
> #11     UpdateStatement._performQuery (package:drift/src/runtime/query_builder/statements/update.dart:36:18)
> <asynchronous suspension>
> #12     UpdateStatement.write (package:drift/src/runtime/query_builder/statements/update.dart:75:12)
> <asynchronous suspension>

Basically, the query to update a int (nullable) column as `null` throws an exception and it doesn't happen with a text nullable column.

Right now I'll have to check if the value to update is null (wherever the type is `int?`) and update as `customUpdate` but most important is not use the `null` parameter in `variables`.
```dart
  db.customUpdate(
    'UPDATE "todos" '
    r'SET "a_int_nullable_column" = $1 '
    r'WHERE "id" = $2;',
    variables: [Variable(null), Variable(1)],
  ); // THROWS SAME ERROR
```

```dart
db.customUpdate(
    'UPDATE "todos" '
    'SET "a_int_nullable_column" = null '
    r'WHERE "id" = $1;',
    variables: [Variable(1)],
  ); // WORKS
```

any help?

**Comment by simolus3** [comment]
Thanks for the report! This is a design limitation in the way `drift` talks to `drift_postgres`. We need to tell postgres that the `null` we're setting `a_int_nullable_column` to is in fact an "integer null", but we don't have that information where we need it because we just see a `null`.

I've worked around the issue in 7810ad9a48f0c17b186959ffb9c2164cce63201f by just spelling out `NULL` in the generated SQL. A proper fix would require a breaking redesign of the low-level database access APIs in drift, I'll keep this issue open as a reminder.

**Comment by simolus3** [comment]
FWIW you can keep using the older `drift_postgres` package that uses the v2 API with newer drift versions too if that worked for you. I don't think anything breaks if you use drift 2.8 from pub but `drift_postgres` 2.5 (or whatever the last version with the old postgres APIs was) from git.
